[Plugin, utilities] Allow writing command output directly to disk

TurboTurtle opened the pull request:
This patchset aims to address a long standing ask for sos - that command output is not _required_ to be stored entirely in memory while we are collecting it.

With these commits, when `--all-logs` is used, or if `sizelimit` is zero for specific commands, `sos_get_command_output()` will write the output directly to a temp file rather than storing the output in memory and returning that.

This is so far only exposed via `add_cmd_output()` by a new `to_file` parameter, but this is automatically handled in the cases above for both `add_cmd_output()` and `add_journal()`. This functionality is not exposed via `exec_cmd()` or `collect_cmd_output()`.

After the command output temp file has been copied to the archive, the temp file is removed via the second commit extending `TempFileUtil`.

Resolves: #1506

---
Please place an 'X' inside each '[]' to confirm you adhere to our [Contributor Guidelines](https://github.com/sosreport/sos/wiki/Contribution-Guidelines)

- [x] Is the commit message split over multiple lines and hard-wrapped at 72 characters?
- [x] Is the subject and message clear and concise?
- [x] Does the subject start with **[plugin_name]** if submitting a plugin patch or a **[section_name]** if part of the core sosreport code?
- [x] Does the commit contain a **Signed-off-by: First Lastname <email@example.com>**?
- [x] Are any related Issues or existing PRs [properly referenced](https://docs.github.com/en/issues/tracking-your-work-with-issues/creating-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword) via a Closes (Issue) or Resolved (PR) line?

TurboTurtle commented:
@pmoravec cycling back around on this.

Do we want to pursue a call to `Archive._check_path()` outside of `add_file()`, or are we ok with the copy-then-delete approach? I'm personally still torn on this.

pmoravec commented:
First, I did some performance measurements and I really appreciate the results: 1GB big command output collected in 10 seconds instead of 30, and with basically no extra memory needed (compared to extra 2GB as now).

I agree size limiting is a blocker to this approach, so I take back it.

About copy&delete vs. populate-in-final-destination: isn't it "just" a matter of careful reordering (and putting under few "`if to_file`" tests) `_collect_cmd_output` ? Now, we generate `outfn` after the command is executed and its output is kept in tempfile or in memory. But the `outfn` relies on command execution outcome only if:

- command not found or runnable or chroot-ed: if we create the leading path in advance, we might end up with empty subdirs - imho no big deal
- result is truncated: this can't happen (as applied to sizelimit=0 only)

Or did I miss something?

TurboTurtle commented:
One concern I had was that we could end up creating empty _files_ for the collections, which may confuse end users into thinking the command was successfully run but produced no output. However, we can always just remove that file if we have no output to write.

I'll work on something for this this week.

TurboTurtle commented:
Sorry for the long silence on this, I wasn't able to dig back into this until now with the other work going on in the project.

Anyways, I've changed this up to write directly to the archive for when `sizelimit=0` or when `--all-logs` is used. From my tests, these scenarios are greatly improved in terms of performance and resource consumption - and "normal" runs are left unaffected.

I dropped the changes to `TempFileUtil` as they were no longer necessary. I think I may also see a way forward for dropping our use of `timeout` for all commands, but that requires more investigation and is out of scope for this particular PR.

pmoravec commented on sos/utilities.py:
```diff
@@ -124,6 +125,12 @@ def _child_prep_fn():
+        if poller():
+            proc.terminate()
+            raise SoSTimeoutError
+        time.sleep(0.1)
```
We had `time.sleep(0.01)`, some reason to increase this? (but I think even the increase does not matter).

TurboTurtle replied:
I ran a couple tests with this while re-writing this section. The TL;DR of it is using `0.1` gives us lower CPU consumption, however at the cost of runtime. This makes sense as we're doing 10x less calls to `poller()`, however we stack up the extra fractions of a second across all the hundreds/thousands of commands we run.

Initially, my opinion is/was that this is preferable - however, I'm now thinking of this in the context of nodes with extremely high numbers of namespaces - which would definitely cause the runtime to increase dramatically more than "normal" systems.

Reverting this back to 0.01 for now, and I'll make a note to do some more investigation on this.

pmoravec reviewed (approved):
Passed all my smoke tests. Performance gain isnt such such great as previously (but I tested it on smaller setup), but still evident improvement.

Great!
